packages/shared-ui/components/cart/reducer/cartReducer.ts

```diff
@@ -0,0 +1,54 @@
+import { CartItem } from "../CartContext";
+
+export type CartUpdate = Partial<CartItem> & {
+  _id: string;
+  quantity: number;
+};
+
+type Action =
+  | { type: "loading" }
+  | { type: "update"; payload: CartUpdate }
+  | { type: "reset" }
+  | { type: "success"; payload: { results: CartItem[]; errors?: string[] } }
+  | { type: "toggleCartOpen"; payload: boolean };
+
+type CartState = {
+  cartItems: CartItem[];
+  state: "loading" | "success";
+  isCartOpen: boolean;
+};
+
+export const initialState: CartState = {
+  cartItems: [],
+  state: "loading",
+  isCartOpen: false,
+};
+
+export const cartReducer = (state: CartState, action: Action): CartState => {
+  switch (action.type) {
+    case "loading":
+      return { ...state, state: "loading" };
+    case "update": {
+      const updateIndex = state.cartItems.findIndex(({ _id }) => _id === action.payload._id);
+      const newCartItems = [
+        ...state.cartItems.slice(0, updateIndex),
+        ...(action.payload.quantity === 0
+          ? []
+          : [
+              {
+                ...state.cartItems[updateIndex],
+                ...action.payload,
+              },
+            ]),
+        ...state.cartItems.slice(updateIndex + 1),
+      ];
+      return { ...state, cartItems: newCartItems };
+    }
```

Comment:
Only question here is should we be calculating (and returning) `totalQuantity, totalPrice` as a part of this reducer when the cart is updated here instead of creating another listener in a `useEffect` in the provider? Trying to think of the reasons we'd want to split up the logic.

Reply:
Good question 👍 I went back and forth on where to put it since it's derived, but I do like the idea of colocating it! I updated the reducer